---
title: "RNAseqPLSRAnalysis"
author: "IM"
date: "2025-03-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#load libraries
library(readr)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(gprofiler2)
library(eulerr)
library(ComplexHeatmap)
library(ggpubr)
library(RColorBrewer)
library(rlang)
library(circlize)
library(pls)
library(Seurat)
library(gprofiler2)
library(eulerr)
library(ComplexHeatmap)
library(ggpmisc)
library(clusterProfiler)
library(org.Hs.eg.db)

#set ligand graphical parameters
combo_cols<- c(
  "EGF+OSM" = "violet",
  "OSM" = "dodgerblue",
  "EGF" = "red",
  "EGF+TGFB" = "orange",
  "OSM+TGFB" = "green4",
  "TGFB" = "yellow2",
  "EGF+OSM+TGFB" = "black",
  "PBS" = "gray65")

```

```{r Functions, echo=FALSE}
#define functions for performing PLSR regression on RNAseq and phenotypic data
#raw and processed RNAseq data are available at the Gene Expression Omnibus: GSE282654
#raw phenotype data is available at 10.5281/zenodo.14261795

#Function to prep RNAseq counts and phenotypic data for PLSR analysis
prep_counts<-function(counts=counts_file, tpm=tpm_file ,phenotype=plsr_y_variables){
  
  #read counts file into Seurat for VST variable feature selection
  counts_seurat_mat<-CreateSeuratObject(counts_file[-c(1:3)])
  counts_seurat_mat<-FindVariableFeatures(counts_seurat_mat, nfeatures = 2500)
  
  #select top 2000 variable features
  variablefeatures2000_lab<-counts_file[counts_seurat_mat@assays$RNA@meta.data$var.features,] %>%
    drop_na((ensembl_gene_id))
  
  #filter for most variant genes
  tpm_rep_var<-tpm_file %>%
    dplyr::select(-gene_biotype) %>%
    filter(ensembl_gene_id %in% variablefeatures2000_lab$ensembl_gene_id) %>%
    select(-hgnc_symbol) %>%
    drop_na() 
  
  #format TPM data and add metadata
  tpm_var<-tpm_rep_var %>%
    select(-ensembl_gene_id) %>%
    t %>%
    as.data.frame() %>%
    rownames_to_column('Name') 

  colnames(tpm_var)<-c('Name',make.names((tpm_rep_var$ensembl_gene_id), unique = T))

  
  Condition = rep(c('CTRL', 'EGF_OSM', 'EGF', 'EGF_TGFB', 'OSM', 'EGF_OSM_TGFB', 'OSM_TGFB', 'PBS',
                  'TGFB'), 3)

  tpm_var$Treatment <-Condition

  Replicate = c(rep(1,9), rep(2,9), rep(3,9))

  tpm_var$Replicate<-Replicate
  
  #add identifier to samples
  tpm_var$SID<-paste0(tpm_var$Treatment, tpm_var$Replicate)
  plsr_y_variables$SID<-paste0(plsr_y_variables$Treatment, plsr_y_variables$Replicate)

  #Normalize tpm file to T0
  all_time_normalized<-data.frame()
  for(replicate in 1:3){
  tpm_var_ctrl <- tpm_var %>%
   filter(Treatment == 'CTRL') 

  tpm_time_normalized_1<-tpm_var %>%
    filter(Replicate == replicate)
  
  t1_ctrl_vector<-as.numeric(tpm_var_ctrl[replicate,c(-1,-2504,-2503, -2502)])
  t1_matrix<-as.matrix(tpm_time_normalized_1[c(-1,-2504,-2503, -2502)])

  tpm_time_normalized_1_b = sweep(as.matrix( t1_matrix), 2, as.numeric(t1_ctrl_vector), '-')

  tpm_time_normalized_1_c<-cbind(tpm_time_normalized_1[c(1,2504,2503, 2502)],tpm_time_normalized_1_b)
  tpm_time_normalized_1_c$Replicate<-replicate
  
  all_time_normalized<-rbind(tpm_time_normalized_1_c, all_time_normalized)%>%
  filter(Treatment != 'CTRL')
  }

  #combine time normalized TPM file and phenotypic metrics to form matrices for PLSR
  combo_combined<-full_join(plsr_y_variables, all_time_normalized)
  combo_X<-as.matrix(combo_combined[-c(1:10)])
  
  return(list(combo_combined, combo_X))
  
}


#evaluate VIP score in PLSR Models to assess gene importance in the model
VIP <- function(object) {
  if (object$method != "oscorespls")
    stop("Only implemented for orthogonal scores algorithm.  Refit with 'method = \"oscorespls\"'")
  if (nrow(object$Yloadings) > 1)
    stop("Only implemented for single-response models")
  
  SS <- c(object$Yloadings)^2 * colSums(object$scores^2)
  Wnorm2 <- colSums(object$loading.weights^2)
  SSW <- sweep(object$loading.weights^2, 2, SS / Wnorm2, "*")
  sqrt(nrow(SSW) * apply(SSW, 1, cumsum) / cumsum(SS))
}

#perform GSEA on top PLSR signatures
GSEA_list<-function(signature){
  
  #select top 100 negatively and positively correlated genes
  vip_format<-signature %>%
    select(ensembl_gene_id, contains('.1 comps'))
  
  colnames(vip_format)<-c('ensembl_gene_id','coefficient')
  
  vip_top <-vip_format %>%
  filter(coefficient>0) %>%
  slice(1:100)
  
  vip_neg <-vip_format %>%
  filter(coefficient<0) %>%
  slice(1:100)
  
  #GSEA of positive and negatively correlated genes
  positive_gsea_results_gp <- gost(query = vip_top$ensembl_gene_id, 
                                 organism = "hsapiens", ordered_query = TRUE, 
                                 multi_query = FALSE, significant = TRUE, 
                                 user_threshold = 0.05, correction_method = "g_SCS", 
                                 numeric_ns = "", as_short_link = FALSE, highlight = TRUE, sources = 'GO')

  top_pathways<-positive_gsea_results_gp$result %>%
    arrange(p_value)%>%
    mutate(logpscore = -log(p_value, base=10))
  
  down_gsea_results_gp <- gost(query = vip_neg$ensembl_gene_id, 
                                    organism = "hsapiens", ordered_query = TRUE, 
                                    multi_query = FALSE, significant = TRUE, 
                                    user_threshold = 0.05, correction_method = "g_SCS", 
                                    numeric_ns = "", as_short_link = FALSE, highlight = TRUE, sources = 'GO')

  top_neg_pathways<-down_gsea_results_gp$result %>%
    arrange(p_value)%>%
    mutate(logpscore = -log(p_value, base=10))
  
  #output to vip lists and pathways
  vip_top<-rbind(vip_top, vip_neg)
  
  return(list(top_pathways, top_neg_pathways, vip_top))
}


```

```{r Data, echo=FALSE}
#import phenotypic features from live cell imaging
#raw data is available at 10.5281/zenodo.14261795, summarised metrics available on github repository
plsr_y_variables<-read_csv( 'Data/Phenotypic_Metrics/Phenotypic_Metrics.csv')

#import replicate raw counts data and tpm file
#count data is available at the Gene Expression Omnibus: GSE282654, TPM file available on github repository
counts_file<-read_csv('GSE282654_MDD_ligandCombination_RNAseq_counts_allGene.csv')
tpm_file<-read_csv('Data/RNAseq_Supp_Files/MDD_ligandCombination_RNAseq_log2TPM_allGenes.csv')

#format TPM file and create PLSR variable matrices
plsr_combined<-prep_counts(counts = counts_file, tpm = tpm_file, phenotype = plsr_y_variables)
combo_combined<-plsr_combined[[1]]
combo_X<-plsr_combined[[2]]

#read in BCCL RNAseq external dataset, raw data available at https://doi.org/10.1073/pnas.1018854108, processed data in github respository
gray_rna<-read_csv('Data/RNAseq_Supp_Files/JWGray_BCCL_rnaseq_matrix_v3_tatlowGeneAbundance.csv')

#read in BCCL phenotypic external raw data https://doi.org/10.1038/s41598-019-47440-w. Processed data on github repository
migration_names<-read_csv('Data/RNAseq_Supp_Files/Migration_BCCL.csv')
prolif_names<-read_csv('Data/RNAseq_Supp_Files/Proliferation_BCCL.csv')

#DEPMAP gene effect data from breast cancer cell lines
#https://doi.org/https://doi.org/10.25452/figshare.plus.25880521.v1
crispr_gene_effect<-read_csv('CRISPRGeneEffect.csv')
models<-read_csv('Model.csv')
```

```{r PLSR}
#cell count PLSR
count_plsr<-plsr(combo_combined$Replicate_Cell_Count~combo_X, ncomp=2, validation='LOO', method="oscorespls", scale=T)

#select components based on RMSEP of fit
plot(RMSEP(count_plsr), legendpos = "topright")
plot(count_plsr, ncomp = 2, asp = 1, line = TRUE)

#extract coefficients from model
count_coeff<-as.data.frame(count_plsr$coefficients) %>%
  rownames_to_column('ensembl_gene_id')

#evaluate fit
cell_count_y_var<-drop(R2(count_plsr, estimate = "train", intercept = FALSE)$val)

#calculate VIP score for whole model weighted by variance explained
cell_count_vip<-as.data.frame(t(VIP(count_plsr))) %>%
  mutate(VIP_SCORE = (`Comp 1`*cell_count_y_var[1] + `Comp 2`*(cell_count_y_var[2] - cell_count_y_var[1]))) %>%
  rownames_to_column('ensembl_gene_id') %>%
  left_join(count_coeff) %>%
  arrange((desc(VIP_SCORE)))

#PLSR model for motility
combo_X<-as.matrix(combo_combined[-c(1:10)])
combo_Y_slope<-scale(as.matrix(combo_combined[c(6)]))

slope_plsr<-plsr(combo_Y_slope~combo_X, ncomp=3, validation='LOO', method="oscorespls", scale=T)

#select components based on RMSEP of fit
plot(RMSEP(slope_plsr), legendpos = "topright")
plot(slope_plsr, ncomp = 3, asp = 1, line = TRUE)

#extract coefficients from model
slope_vip<-as.data.frame(t(VIP(slope_plsr))) %>%
  arrange(desc(`Comp 1`)) %>%
  rownames_to_column('ensembl_gene_id')
slope_coeff<-as.data.frame(slope_plsr$coefficients) %>%
  rownames_to_column('ensembl_gene_id')

#evaluate fit
slope_y_var<-drop(R2(slope_plsr, estimate = "train", intercept = FALSE)$val)

#calculate VIP score for whole model weighted by variance explained
slope_top<-slope_vip %>%
  mutate(VIP_SCORE =(`Comp 1`*slope_y_var[1]  + `Comp 2`*(slope_y_var[2]-slope_y_var[1]) + `Comp 3`*(slope_y_var[3] - slope_y_var[2]))) %>%
  left_join(slope_coeff) %>%
  arrange(desc(VIP_SCORE))

#PLSR model for Nearest Neighbor
combo_X<-as.matrix(combo_combined[-c(1:10)])
combo_Y_neighbor<-scale(as.matrix(combo_combined[c(7)]))

neighbor_plsr<-plsr(combo_Y_neighbor~combo_X, ncomp=2, validation='LOO', method="oscorespls", scale=T)

#select components based on RMSEP
plot(RMSEP(neighbor_plsr), legendpos = "topright")
plot(neighbor_plsr, ncomp = 2, asp = 1, line = TRUE)

#extract coefficiennts and VIP
neighbor_vip<-as.data.frame(t(VIP(neighbor_plsr))) %>%
  arrange(desc(`Comp 1`)) %>%
  rownames_to_column('ensembl_gene_id')

neighbor_coeff<-as.data.frame(neighbor_plsr$coefficients) %>%
  rownames_to_column('ensembl_gene_id')

#evaluate fit
neighbor_y_var<-drop(R2(neighbor_plsr, estimate = "train", intercept = FALSE)$val)

#calculate weighted VIP scores
neighbor_top<-neighbor_vip %>%
  mutate(VIP_SCORE =(`Comp 1`*neighbor_y_var[1]  + `Comp 2`*(neighbor_y_var[2]-neighbor_y_var[1]))) %>%
  arrange(desc(VIP_SCORE)) %>%
  left_join(neighbor_coeff) 

#PLSR model for cytoplasmic size
combo_X<-as.matrix(combo_combined[-c(1:10)])
combo_Y_cytoplasm<-scale(as.matrix(combo_combined[c(8)]))

cytoplasm_plsr<-plsr(combo_Y_cytoplasm~combo_X, ncomp=5, validation='LOO', method="oscorespls", scale=T)

#select components based on RMSEP
plot(RMSEP(cytoplasm_plsr), legendpos = "topright")
plot(cytoplasm_plsr, ncomp = 5, asp = 1, line = TRUE)

#extract VIP and coefficients
cytoplasm_vip<-as.data.frame(t(VIP(cytoplasm_plsr))) %>%
  arrange(desc(`Comp 1`)) %>%
  rownames_to_column('ensembl_gene_id')

cytoplasm_coeff<-as.data.frame(cytoplasm_plsr$coefficients) %>%
  rownames_to_column('ensembl_gene_id')

#evaluate fit
cytoplasm_y_var<-drop(R2(cytoplasm_plsr, estimate = "train", intercept = FALSE)$val)

#calculate weighted VIP scores
cytoplasm_top<-cytoplasm_vip %>%
  mutate(VIP_SCORE =(`Comp 1`*cytoplasm_y_var[1]  + `Comp 2`*(cytoplasm_y_var[2] - cytoplasm_y_var[1]) +  `Comp 3`*(cytoplasm_y_var[3] - cytoplasm_y_var[2]) +
                       `Comp 4`*(cytoplasm_y_var[4] - cytoplasm_y_var[3]) + `Comp 5`*(cytoplasm_y_var[5] - cytoplasm_y_var[4]))) %>%
  left_join(cytoplasm_coeff) %>%
  arrange(desc(VIP_SCORE))
```


```{r GSEA analysis of phenotypic gene signatures}
#Perform overrepresentation GSEA on top 100 and bottom 100 genes in PLSR model correlated to phenotype
count_gsea<-GSEA_list(cell_count_vip)
motility_gsea<-GSEA_list(slope_top)
nearestneighbor_gsea<-GSEA_list(neighbor_top)
cytoplasm_gsea<-GSEA_list(cytoplasm_top)

#barplot of top pathways
ggplot(count_gsea[[1]][1:10,], aes(x = reorder(term_name, (logpscore)), y = logpscore)) + 
  geom_bar(stat = 'identity') + coord_flip()  + scale_y_reverse(limits = c(8,0)) + scale_x_discrete(labels=function(x) str_wrap(x, width=25)) +  labs(x = "log10(q-value)") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 0, size=rel(2)),
        axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=rel(2)),
        plot.title = element_text(size = rel(2)),legend.text=element_text(size = rel(1)), axis.title.y = element_blank(),axis.title = element_text(size = rel(2)),
        legend.title=element_text(size = rel(2)), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggplot(count_gsea[[2]][1:10,], aes(x = reorder(term_name, (logpscore)), y = logpscore)) + 
  geom_bar(stat = 'identity') + coord_flip()  + scale_x_discrete(labels=function(x) str_wrap(x, width=25), position = 'top') +  labs(x = "log10(q-value)") + scale_y_continuous(limits = c(0,8)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 0, size=rel(2)),
        axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=rel(2)),
        plot.title = element_text(size = rel(2)),legend.text=element_text(size = rel(1)), axis.title.y = element_blank(), axis.title = element_text(size = rel(2)),
        legend.title=element_text(size = rel(2)), panel.grid.major = element_blank(), panel.grid.minor = element_blank())

#heatmap of gene signature
#make heatmaps of top 100 positively and negatively correlated VIPS
combo_combined_hm<-combo_combined %>%
  dplyr::select(Treatment, Replicate, Replicate_Cell_Count, count_gsea[[3]]$ensembl_gene_id) %>%
  arrange(desc(Replicate_Cell_Count))

vip_hm<-as.matrix(t(combo_combined_hm[-c(1:3)])) %>% t %>%
  scale() %>%
  t
colnames(vip_hm)<-combo_combined_hm$Treatment

colnames(vip_hm)<-gsub('_','+',colnames(vip_hm))
ha2 = HeatmapAnnotation(`Cell Count` = anno_barplot(combo_combined_hm$Replicate_Cell_Count, axis_param=list(gp=gpar(fontsize = 16))), height = unit(3, "cm"), annotation_name_gp= gpar(fontsize = 14, fontface = 'bold'))

ht<-Heatmap(matrix = vip_hm,
            name = "Z-Score",
            column_title = 'Top PLSR Correlates to Cytoplasmic Size',
            show_row_names = F,
            show_column_names = T,
            show_row_dend = F,
            show_column_dend = F,
            cluster_columns = F,
            cluster_rows = F,
            row_title_gp = gpar(fontsize = 22),
            top_annotation = ha2,
            row_split = factor(append(rep(("Positive"), 100), rep('Negative',100)), levels = c('Positive','Negative')),
            heatmap_legend_param = list(labels_gp = gpar( fontsize = 18), title_gp = gpar( fontsize = 18)),
            column_title_gp = gpar(fontsize = 23),
            column_names_gp = gpar(fontsize = 16),
            row_names_gp = gpar(fontsize = 16))
```

```{r Model external BCCL datasets using PLSR model}
#predict migration and proliferation with entire PLSR model
#format BCCL data to agree with bulk RNAseq format and so that genes agree and cell lines are consistent, and scale expression data
gray_rna_log2<-gray_rna %>%
  dplyr::rename(ensembl_gene_id = gene_ensg) %>%
  mutate(ensembl_gene_id = gsub("\\..*",'',ensembl_gene_id)) %>%
  column_to_rownames('ensembl_gene_id') %>%
  mutate(across(where(is.numeric), ~ .x + 1)) %>%
  mutate(across(where(is.numeric), ~ log2(.x)))

filtered_mapped_matrix <- gray_rna_log2 %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column('Cell_Line') %>%
  mutate(Cell_Line = gsub("_.*",'',Cell_Line)) %>%
  column_to_rownames('Cell_Line') %>%
  t

scaled_gray<-filtered_mapped_matrix %>%
  as.data.frame() %>%
  rownames_to_column('ensembl_gene_id') %>%
  filter(ensembl_gene_id %in% colnames(combo_X)) %>%
  column_to_rownames('ensembl_gene_id') %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column('Cell_Line') %>%
  filter(Cell_Line %in% migration_names$Cell_Line) %>%
  mutate_if(is.numeric,scale) 

scaled_gray[is.na(scaled_gray)] <- 0

scaled_gray_form <-scaled_gray %>%
  column_to_rownames('Cell_Line') %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column('ensembl_gene_id') %>%
  filter(ensembl_gene_id %in% colnames(combo_X)) %>%
  column_to_rownames('ensembl_gene_id') %>%
  t

#filter model for genes in scaled_gray
combo_X_trimmed<-(combo_X) %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column('ensembl_gene_id') %>%
  filter(ensembl_gene_id %in% colnames(scaled_gray_form)) %>%
  column_to_rownames('ensembl_gene_id') %>%
  t

combo_Y_count<-scale(as.matrix(combo_combined[c(4)]))

#predict motility and cell count
count_plsr<-plsr(combo_Y_count~combo_X_trimmed, ncomp=6, validation='LOO', method="oscorespls", scale=T)
count_prediction<-predict(count_plsr, ncomp = 1, newdata = scaled_gray_form)

#z-score predictions
proliferation_prediction_z<-count_prediction %>%
  as.data.frame() %>%
  rownames_to_column('Cell_Line') %>%
  left_join(prolif_names)

proliferation_prediction_z$proliferation_prediction_zscore = scale(proliferation_prediction_z$`combo_Y_count.1 comps`) 

prolif_mean = mean(as.matrix(combo_combined[c(4)]))
prolif_sd = sd(as.matrix(combo_combined[c(4)]))

proliferation_prediction_z$prolif_prediction = proliferation_prediction_z$`combo_Y_count.1 comps`*prolif_sd + prolif_mean

t<-ggplot(proliferation_prediction_z, aes(x=prolif_prediction, y = Proliferation))+
  geom_point(size = 4) + theme_bw() + coord_flip() + xlab('Predicted Proliferation') + ylab('Experimental Proliferation') +
  stat_poly_line() + scale_y_continuous(breaks = c(.5,1,1.5,2)) +
  stat_poly_eq(use_label(c('R2','P')), size = 8)+
  theme(axis.text.x = element_text(angle = 0, size=rel(2.5)),
        plot.margin = margin(10, 11, 10, 10),
        axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=rel(2.5)),
        plot.title = element_text(size = rel(3)),legend.text=element_text(size = rel(1)), axis.title = element_text(size = rel(3)),
        legend.title=element_text(size = rel(2)),  panel.grid.minor = element_blank())


```

```{r DEPMAP analysis}
#look at correlation between cell count predictor and DEPMAP essentiallity
#format data from depmap
crispr_format<-crispr_gene_effect %>%
  rename(ModelID = ...1) %>%
  column_to_rownames('ModelID') %>%
  t %>%
  as.data.frame() %>%
  rownames_to_column('hgnc_symbol') %>%
  mutate(hgnc_symbol = gsub(' .*', '',hgnc_symbol))
  
crispr_ensembl_ids<-bitr(crispr_format$hgnc_symbol, fromType = "SYMBOL",
                                          toType = c("ENSEMBL"),
                                          OrgDb = org.Hs.eg.db) %>%
  dplyr::rename(hgnc_symbol = SYMBOL) %>%
  dplyr::rename(ensembl_gene_id = ENSEMBL)

crispr_renamed<-left_join(crispr_format, crispr_ensembl_ids)

#filter for breast cancer cell lines
breast_models<-models %>%
  filter(OncotreeLineage == 'Breast')

all_achilles<-crispr_renamed %>%
  filter(ensembl_gene_id %in% cell_count_vip$ensembl_gene_id) %>%
  gather(key = ModelID, value = GeneEffect, -ensembl_gene_id, -hgnc_symbol) %>%
  filter(ModelID %in% breast_models$ModelID) 

#look at essentiality of genes in heatmap for proliferation
all_gene_vip<-left_join(all_achilles, cell_count_vip)

hm_genes<-all_gene_vip %>%
  filter(ensembl_gene_id %in% count_gsea[[3]]$ensembl_gene_id)

#order the genes to be consistent with gene heatmap
order_genes_top<-hm_genes %>%
  rownames_to_column('Order') %>%
  left_join(all_gene_vip) %>%
  filter(`combo_combined$Replicate_Cell_Count.1 comps`>0)
order_genes_top$Order<-as.numeric(order_genes_top$Order)
order_genes_bottom<-hm_genes %>%
  rownames_to_column('Order') %>%
  left_join(all_gene_vip) %>%
  filter(`combo_combined$Replicate_Cell_Count.1 comps`<0)
order_genes_bottom$Order<-as.numeric(order_genes_bottom$Order)

#plot depmap score
top<-ggplot(order_genes_top, aes(x=GeneEffect, y = rev(Order))) +
  geom_point() + xlab('Gene Effect') + ylab('VIP Score') + theme_bw() + geom_vline(xintercept = -1, linetype = 'dashed', color = 'red') +
  theme(axis.text.x = element_text(angle = 0, size=rel(2)),
        plot.margin = margin(10, 11, 10, 10),
        axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
        plot.title = element_text(size = rel(2)),legend.text=element_text(size = rel(1)), axis.title = element_text(size = rel(2)),
        legend.title=element_text(size = rel(2)), panel.grid.major = element_blank(),  panel.grid.minor = element_blank())
bottom<-ggplot(order_genes_bottom, aes(x=GeneEffect, y = rev(Order))) +
  geom_point() + xlab('Gene Effect') + ylab('VIP Score') + theme_bw() + geom_vline(xintercept = -1, linetype = 'dashed', color = 'red') +
  theme(axis.text.x = element_text(angle = 0, size=rel(2)),
        plot.margin = margin(10, 11, 10, 10),
        axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
        plot.title = element_text(size = rel(2)),legend.text=element_text(size = rel(1)), axis.title = element_text(size = rel(2)),
        legend.title=element_text(size = rel(2)), panel.grid.major = element_blank(),  panel.grid.minor = element_blank())

top+bottom

#run chi squared on Geneeffect <= -1 and VIP score >=1
all_gene_vip_hi_lo<-all_gene_vip %>%
  mutate(gene_effect_direction = ifelse(GeneEffect <= -.5, 'Essential_Gene',
         'NonEssential_Gene')) %>%
  mutate(VIP_direction = ifelse(VIP_SCORE >= 1, 'High_VIP',
                                        'Low_VIP'))

chisq.test(all_gene_vip_hi_lo$gene_effect_direction, all_gene_vip_hi_lo$VIP_direction, correct=FALSE)

#look at all genes
all_gene_vip_median<-all_gene_vip_hi_lo %>%
  group_by(ensembl_gene_id, VIP_SCORE) %>%
  summarise(GeneEffect = mean(GeneEffect)) %>%
  drop_na()

#color genes by essentiality
all_gene_vip_median$`Mean Value`[all_gene_vip_median$GeneEffect<(-.5)] = '< -0.5'
all_gene_vip_median$`Mean Value`[all_gene_vip_median$GeneEffect>(-.5)] = '> -0.5'

#join genes to cell count signature
all_gene_vip_combined<-all_gene_vip_hi_lo %>%
  dplyr::select(VIP_SCORE, GeneEffect, ensembl_gene_id) %>%
  mutate(`Mean Value` = '') %>%
  full_join(all_gene_vip_median) %>%
  arrange(desc(VIP_SCORE))

#plot essentiality and VIP Score
ggplot(all_gene_vip_combined, aes(x=GeneEffect, y = VIP_SCORE, color = `Mean Value`, size = `Mean Value`, alpha = `Mean Value`)) +
  geom_point() + xlab('Gene Effect') + ylab('VIP Score') + theme_bw() + geom_vline(xintercept = -.5, linetype = 'dashed', color = 'red')+ scale_alpha_manual(values = c("> -0.5" = 1, "< -0.5"=1, .3)) +
  scale_x_reverse(lim =c(0,-5)) + scale_color_manual(values = c("> -0.5" = 'gray39', "< -0.5"="red4", 'gray'), breaks = c("< -0.5", "> -0.5")) +
  scale_size_manual(values = c("< -0.5" = 3, "> -0.5"=3, 1.5)) + guides(color=guide_legend(title="Mean Gene Effect"), size = 'none', alpha = 'none') +
  theme(axis.text.x = element_text(angle = 0, size=rel(2)),strip.text.y = element_blank(),strip.text = element_blank(),
        strip.background = element_blank(),
        plot.margin = margin(10, 11, 10, 10),
        axis.text.y = element_text(size=rel(2)), 
        plot.title = element_text(size = rel(2)),legend.text=element_text(size = rel(1.5)), axis.title = element_text(size = rel(2)),
        legend.title=element_text(size = rel(2)), panel.grid.minor = element_blank())

```

